KAFKA-7793: Improve the Trogdor command line.

cmccabe opened the pull request:
* Allow the Trogdor agent to be started in "exec mode", where it simply
runs a single task and exits after it is complete.

* For AgentClient and CoordinatorClient, allow the user to pass the path
to a file containing JSON, instead of specifying the JSON object in the
command-line text itself.  This means that we can get rid of the bash
scripts whose only function was to load task specs into a bash string
and run a Trogdor command.

* Print dates and times in a human-readable way, rather than as numbers
of milliseconds.

* When listing tasks or workers, output human-readable tables of
information.

* Allow the user to filter on task ID name, task ID pattern, or task
state.

* Support a --json flag to provide raw JSON output if desired.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

stanislavkozlovski commented on tests/spec/round_trip.json:
```diff
@@ -0,0 +1,14 @@
+  "targetMessagesPerSec": 1000,
+  "maxMessages": 100,
+  "activeTopics": {
+    "${TASK_ID}_topic[0-1]": {
```
Do we support templating now?

cmccabe replied:
Oops-- this was an oversight.  Fixed.

stanislavkozlovski commented on tests/spec/round_trip.json:
```diff
@@ -0,0 +1,14 @@
+{
```
I'm wondering whether we should add a `README.md` under `/tests/spec/` or rename to something similar to `/tests/trogdor_specs` - just to make it clearer for new users what these specs represent

cmccabe replied:
Good idea.  I added an Apache header and a comment to the files, referring to `TROGDOR.md`

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/agent/AgentClient.java:
```diff
@@ -150,87 +161,147 @@ public void invokeShutdown() throws Exception {
+        Subparser createWorkerParser = subParsers.addParser("createWorker")
+            .help("Create a new worker.");
+        addTargetArgument(createWorkerParser);
+        createWorkerParser.addArgument("--workerId")
```
nit: Might as well create an `addWorkerIdArgument()` to be consistent with the other shared args

stanislavkozlovski commented on tools/src/test/java/org/apache/kafka/trogdor/common/JsonUtilTest.java:
```diff
@@ -0,0 +1,84 @@
+            try {
+                Files.delete(tempFile.toPath());
+            } catch (IOException e) {
+                log.error("Unable to delete " + tempFile, e);
```
nit: probably not worth it to log an error, as another delete will be attempted.  `TestUtils.tempFile()` calls `file.deleteOnExit();`

stanislavkozlovski commented on tools/src/test/java/org/apache/kafka/trogdor/common/JsonUtilTest.java:
```diff
@@ -0,0 +1,84 @@
+    public void testOpenBraceComesFirst() {
+        assertTrue(JsonUtil.openBraceComesFirst("{}"));
+        assertTrue(JsonUtil.openBraceComesFirst(" \t{\"foo\":\"bar\"}"));
+        assertTrue(JsonUtil.openBraceComesFirst(" \t{\"foo\":\"bar\"}"));
```
Those two test cases are exactly the same, I assume you wanted to test one with spaces (` `) only?

cmccabe replied:
Fixed, thanks

stanislavkozlovski commented on tools/src/test/java/org/apache/kafka/trogdor/coordinator/CoordinatorClientTest.java:
```diff
@@ -0,0 +1,83 @@
+                new TaskRunning(new NoOpTaskSpec(1146931159850L, 9000),
+                    1246931159850L,
+                    JsonNodeFactory.instance.objectNode()), ZoneOffset.UTC));
+        assertEquals("Started 2009-07-07T01:45:59.85Z",
```
Should we print a status here that the task is in the process of stopping?

cmccabe replied:
That shows up in the task state, which is STOPPING.  But the specific function here, `CoordinatorClient#prettyPrintTaskInfo`, doesn't print that.  So I think the test is OK.

stanislavkozlovski commented on TROGDOR.md:
```diff
@@ -35,61 +35,26 @@ Let's confirm that all of the daemons are running:
-
-To get the test results, we run --show-tasks:
-
-    ./bin/trogdor.sh client --show-tasks localhost:8889
```
To clarify, this behavior of printing out the full JSON is preserved if we were to send `--json`, right?

cmccabe replied:
yes

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/agent/Agent.java:
```diff
@@ -106,6 +141,58 @@ public void destroyWorker(DestroyWorkerRequest req) throws Throwable {
         workerManager.stopWorker(req.workerId(), true);
     }
 
+    TaskSpec rebaseTaskSpecTime(TaskSpec spec) throws Exception {
```
I guess it shouldn't have any big implications but maybe it's worth noting in the java doc that this is only done in `exec` mode and in normal circumstances we use the `startMs` provided by the coordinator?

cmccabe replied:
Yeah.  Added Javadoc.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/agent/WorkerManager.java:
```diff
@@ -345,6 +352,7 @@ public void createWorker(long workerId, String taskId, TaskSpec spec) throws Thr
                     "worker.start() exception: " + Utils.stackTrace(e), true));
             }
             stateChangeExecutor.submit(new FinishCreatingWorker(worker));
+            return worker.doneFuture;
```
If we execute the `haltFuture` path above of (line: 343)
```
stateChangeExecutor.submit(new HandleWorkerHalting(worker, errorString, false));
return null;
```
I don't think we will ever call `transitionToDone()`, therefore leaving the `--exec` call hanging for some time.

I guess we will once the task expires via:
```
void transitionToRunning() {
            state = State.RUNNING;
            timeoutFuture = scheduler.schedule(stateChangeExecutor,
                new StopWorker(workerId, false),
                Math.max(0, spec.endMs() - time.milliseconds()));
        }
```
but it would be better to have it be faster and work on tasks with huge duration times.

Could you verify this?

cmccabe replied:
I don't think this is accurate.  `HandleWorkerHalting` is supposed to get the worker to `DONE`.  For example, if the problem is that the worker halted during startup, it just calls `transitionToDone` directly.  If the worker halted after startup, then it will have to go through other states like `CANCELLING` or `STOPPING` first, but it will arrive at `DONE` eventually.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
+            .help("Show a coordinator task.");
+        addTargetArgument(showTaskParser);
+        addJsonArgument(showTaskParser);
+        showTaskParser .addArgument("--id", "-i")
```
nit: extra space in between `showTaskParser` and `.addArgument`

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
-            .help("Show a specific coordinator task.");
-        actions.addArgument("--create-task")
+            .help("The task ID to show.");
+        showTaskParser .addArgument("--verbose", "-v")
```
nit: extra space

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
+            .dest("verbose")
+            .metavar("VERBOSE")
+            .help("Print out everything.");
+        showTaskParser .addArgument("--show-status", "-S")
```
nit: extra space

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
+            .dest("taskIdPattern")
+            .metavar("TASK_ID_PATTERN")
+            .help("Only display tasks which match the given ID pattern.");
+        showTasksParser.addArgument("--state")
```
Could we add a shorthand `--s` or chance `--show-status`'s shorthand to `--s`?

cmccabe replied:
Added `-s` as shorthand for `--state`

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
+                }
+                TasksRequest req = new TasksRequest(taskIds, 0, 0, 0, 0,
+                    Optional.ofNullable(taskStateType));
+                TasksResponse response = client.tasks(req);
```
Will this work with multiple task ids? The `tasks()` code does this:
```
        uriBuilder.queryParam("taskId", (Object[]) request.taskIds().toArray(new String[0]));
```
I think that returns a single-element array only?

cmccabe replied:
`List#toArray` can return arrays with more than one element.

https://docs.oracle.com/javase/8/docs/api/java/util/List.html#toArray-T:A-

The array argument at the end is a little weird.  TL;DR is that if you want to squeeze out some extra efficiency, you can pass something other than a zero-length array to that argument, to avoid object allocation.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/coordinator/CoordinatorClient.java:
```diff
@@ -170,111 +193,291 @@ public void shutdown() throws Exception {
+                if (taskIdPattern != null) {
+                    TreeMap<String, TaskState> filteredTasks = new TreeMap<>();
+                    for (Map.Entry<String, TaskState> entry : response.tasks().entrySet()) {
+                        if (taskIdPattern.matcher(entry.getKey()).matches()) {
```
This is great functionality but would it be better if we had it in the coordinator server in order to leverage it from REST API clients in other languages? Otherwise we probably need to duplicate logic.
 We could maybe open a JIRA to track this as a separate item

cmccabe replied:
We've been holding off on including regular expressions into APIs since we would like APIs to be language neutral, and Java regexps are not.  They also change over time as Java changes (java 8 changed how some of them are interpreted).

I think we could open a follow-up JIRA for this, though.  We can use re2 regular expressions to make the API language-neutral.

stanislavkozlovski reviewed (commented):
Lots of great changes! Thanks for the PR, @cmccabe!

I like the separation into `.json` specs a lot - it is more friendly for an abundance of examples rather than all the `.sh` scripts we had to duplicate logic in.

cmccabe committed:
Fixes to spec files

cmccabe committed:
KAFKA-7793: Improve the Trogdor command line.

* Allow the Trogdor agent to be started in "exec mode", where it simply
runs a single task and exits after it is complete.

* For AgentClient and CoordinatorClient, allow the user to pass the path
to a file containing JSON, instead of specifying the JSON object in the
command-line text itself.  This means that we can get rid of the bash
scripts whose only function was to load task specs into a bash string
and run a Trogdor command.

* Print dates and times in a human-readable way, rather than as numbers
of milliseconds.

* When listing tasks or workers, output human-readable tables of
information.

* Allow the user to filter on task ID name, task ID pattern, or task
state.

* Support a --json flag to provide raw JSON output if desired.

cmccabe committed:
Address review comments

mumrah commented on tools/src/main/java/org/apache/kafka/trogdor/common/JsonUtil.java:
```diff
@@ -53,4 +57,41 @@ public static String toPrettyJsonString(Object object) {
     }
+
+    /**
+     * Determine if a string is a JSON object literal.
```
Do we have any use cases for lists of objects at the top level of a JSON string?

cmccabe replied:
That's a fair question.  The answer here is that in all the cases where we're using this function, we want a JSON object as the top-level thing.  We generally always make things objects in case we need to add more stuff later on.  (For example, perhaps right now we want to configure X with a list of ints, but later we might want a list of ints and a float, etc.)

mumrah commented on tools/src/main/java/org/apache/kafka/trogdor/common/Formatter.java:
```diff
@@ -0,0 +1,114 @@
+/**
+ * Utilities for formatting strings.
+ */
+public class Formatter {
```
Maybe `StringFormatter`?

cmccabe replied:
Sounds good.  Renamed

mumrah reviewed (commented):
I don't see any examples of `--exec` usage in the README. Maybe we could add one?

Supplying the spec JSON in a file rather than having to do shell gymnastics is definitely an improvement. Human readable output is always nice (for us humans, at least).

cmccabe committed:
Formatter -> StringFormatter

cmccabe committed:
Add exec mode to TROGDOR.md

cmccabe commented:
retest this please

mumrah reviewed (approved):
LGTM

stanislavkozlovski reviewed (approved):
JDK8 build test failure is unrelated - `kafka.api.AdminClientIntegrationTest.testForceClose`

LGTM! Working with Trogdor locally just got a heck of a lot better - thanks a lot for these changes @cmccabe!
